File: clean_architecture/data_generation.py
from cassandra.cluster import Cluster
from uuid import uuid4
import random
import string

# Connect to the Cassandra cluster
cluster = Cluster(['localhost'])
session = cluster.connect()

# Create keyspace and use it
session.execute("""
CREATE KEYSPACE IF NOT EXISTS tagging 
WITH REPLICATION = { 'class' : 'SimpleStrategy', 'replication_factor' : 1 };
""")
session.set_keyspace('tagging')

# Create tables
session.execute("""
CREATE TABLE IF NOT EXISTS tags (
    id UUID PRIMARY KEY,
    owner_id INT,
    name TEXT,
    description TEXT,
    properties TEXT
);
""")

session.execute("""
CREATE TABLE IF NOT EXISTS entities (
    id UUID PRIMARY KEY,
    name TEXT,
    type TEXT,
    metadata TEXT
);
""")

session.execute("""
CREATE TABLE IF NOT EXISTS tag_entities (
    tag_id UUID,
    entity_id UUID,
    PRIMARY KEY (tag_id, entity_id)
);
""")

# Generate random data
def random_string(length=10):
    letters = string.ascii_letters
    return ''.join(random.choice(letters) for i in range(length))

# Insert records into the 'tags' table
tags = []
for _ in range(1000000):
    tag_id = uuid4()
    tags.append(tag_id)
    session.execute(
        """
        INSERT INTO tags (id, owner_id, name, description, properties)
        VALUES (%s, %s, %s, %s, %s)
        """,
        (tag_id, random.randint(1, 1000), random_string(15), random_string(50), random_string(100))
    )

print("Inserted 1 million records into 'tags'")

# Insert records into the 'entities' table
entities = []
for _ in range(1000000):
    entity_id = uuid4()
    entities.append(entity_id)
    session.execute(
        """
        INSERT INTO entities (id, name, type, metadata)
        VALUES (%s, %s, %s, %s)
        """,
        (entity_id, random_string(15), random_string(10), random_string(100))
    )

print("Inserted 1 million records into 'entities'")

# Insert records into the 'tag_entities' table using IDs from 'tags' and 'entities'
batch_size = 1000  # Number of records to insert in a batch
for i in range(0, 1000000, batch_size):
    batch = []
    for j in range(batch_size):
        tag_id = tags[random.randint(0, len(tags) - 1)]
        entity_id = entities[random.randint(0, len(entities) - 1)]
        batch.append((tag_id, entity_id))
    prepared = session.prepare("INSERT INTO tag_entities (tag_id, entity_id) VALUES (?, ?)")
    for record in batch:
        session.execute(prepared, record)
    print(f"Inserted {i + batch_size} records into 'tag_entities'")

# Close the connection
cluster.shutdown()

File: clean_architecture/data_generation.go
package main

import (
	"fmt"
	"github.com/gocql/gocql"
	"github.com/google/uuid"
	"log"
	"math/rand"
	"time"
)

func randomString(length int) string {
	chars := []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
	s := make([]rune, length)
	for i := range s {
		s[i] = chars[rand.Intn(len(chars))]
	}
	return string(s)
}

func main() {
	// Set up the cluster configuration
	cluster := gocql.NewCluster("127.0.0.1")
	cluster.Keyspace = "tagging"
	cluster.Consistency = gocql.Quorum

	// Connect to the cluster
	session, err := cluster.CreateSession()
	if err != nil {
		log.Fatal(err)
	}
	defer session.Close()

	// Create keyspace and use it
	if err := session.Query(`CREATE KEYSPACE IF NOT EXISTS tagging WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1}`).Exec(); err != nil {
		log.Fatal(err)
	}

	// Create tables
	if err := session.Query(`CREATE TABLE IF NOT EXISTS tags (
		id UUID PRIMARY KEY,
		owner_id INT,
		name TEXT,
		description TEXT,
		properties TEXT
	)`).Exec(); err != nil {
		log.Fatal(err)
	}

	if err := session.Query(`CREATE TABLE IF NOT EXISTS entities (
		id UUID PRIMARY KEY,
		name TEXT,
		type TEXT,
		metadata TEXT
	)`).Exec(); err != nil {
		log.Fatal(err)
	}

	if err := session.Query(`CREATE TABLE IF NOT EXISTS tag_entities (
		tag_id UUID,
		entity_id UUID,
		PRIMARY KEY (tag_id, entity_id)
	)`).Exec(); err != nil {
		log.Fatal(err)
	}

	rand.Seed(time.Now().UnixNano())

	// Generate and insert 1 million records into 'tags' table
	for i := 0; i < 10; i++ {
		tagID, _ := gocql.ParseUUID(uuid.New().String())
		ownerID := rand.Intn(1000) + 1
		name := randomString(15)
		description := randomString(50)
		properties := randomString(100)
		if err := session.Query(`INSERT INTO tags (id, owner_id, name, description, properties) VALUES (?, ?, ?, ?, ?)`,
			tagID, ownerID, name, description, properties).Exec(); err != nil {
			log.Fatal(err)
		}
		if i%1000 == 0 {
			fmt.Printf("Inserted %d records into 'tags'\n", i+1)
		}
	}

	// Generate and insert 1 million records into 'entities' table
	for i := 0; i < 10; i++ {
		entityID, _ := gocql.ParseUUID(uuid.New().String())
		name := randomString(15)
		entityType := randomString(10)
		metadata := randomString(100)
		if err := session.Query(`INSERT INTO entities (id, name, type, metadata) VALUES (?, ?, ?, ?)`,
			entityID, name, entityType, metadata).Exec(); err != nil {
			log.Fatal(err)
		}
		if i%1000 == 0 {
			fmt.Printf("Inserted %d records into 'entities'\n", i+1)
		}
	}

	// Select all tag IDs and entity IDs to use in the tag_entities table
	tagIDs := []gocql.UUID{}
	entityIDs := []gocql.UUID{}

	iter := session.Query(`SELECT id FROM tags`).Iter()
	var id gocql.UUID
	for iter.Scan(&id) {
		id, _ := gocql.ParseUUID(uuid.New().String())
		tagIDs = append(tagIDs, id)
	}
	if err := iter.Close(); err != nil {
		log.Fatal(err)
	}

	iter = session.Query(`SELECT id FROM entities`).Iter()
	for iter.Scan(&id) {
		id, _ := gocql.ParseUUID(uuid.New().String())
		entityIDs = append(entityIDs, id)
	}
	if err := iter.Close(); err != nil {
		log.Fatal(err)
	}

	// Generate and insert records into 'tag_entities' table using IDs from 'tags' and 'entities'
	for i := 0; i < 10; i++ {
		//entityID,_:= gocql.ParseUUID(uuid.New().String())
		tagID, _ := gocql.ParseUUID(tagIDs[rand.Intn(len(tagIDs))].String())
		entityID, _ := gocql.ParseUUID(entityIDs[rand.Intn(len(entityIDs))].String())

		if err := session.Query(`INSERT INTO tag_entities (tag_id, entity_id) VALUES (?, ?)`,
			tagID, entityID).Exec(); err != nil {
			log.Fatal(err)
		}
		if i%1000 == 0 {
			fmt.Printf("Inserted %d records into 'tag_entities'\n", i+1)
		}
	}

	fmt.Println("Data generation completed.")
}
